File: packages/shell-api2/src/AggregationCursor.ts
import Mongo from './mongo';
import { shellApiClassDefault, returnsPromise, returnType } from './main';
import {
  Cursor as ServiceProviderCursor,
  Document
} from '@mongosh/service-provider-core';

@shellApiClassDefault
export default class AggregationCursor {
  mongo: Mongo;
  cursor: ServiceProviderCursor;
  constructor(mongo, cursor) {
    this.cursor = cursor;
    this.mongo = mongo;
  }

  toReplString() {
    return this.mongo.it();
  }

  @returnsPromise
  close(options: Document): Promise<void> {
    return this.cursor.close(options);
  }

  @returnsPromise
  forEach(f): Promise<void> {
    return this.cursor.forEach(f);
  }

  @returnsPromise
  hasNext(): Promise<boolean> {
    return this.cursor.hasNext();
  }

  isClosed(): boolean {
    return this.cursor.isClosed();
  }

  isExhausted(): Promise<boolean> {
    return this.cursor.isExhausted();
  }

  @returnsPromise
  itcount(): Promise<number> {
    return this.cursor.itcount();
  }

  @returnType('AggregationCursor')
  map(f): AggregationCursor {
    this.cursor.map(f);
    return this;
  }

  @returnsPromise
  next(): Promise<any> {
    return this.cursor.next();
  }

  @returnsPromise
  toArray(): Promise<Document[]> {
    return this.cursor.toArray();
  }

  @returnsPromise
  explain(verbosity: string): Promise<any> {
    return this.cursor.explain(verbosity);
  }
}

File: packages/java-shell/src/main/kotlin/com/mongodb/mongosh/result/AggregationCursor.kt
package com.mongodb.mongosh.result

import com.mongodb.mongosh.MongoShellContext
import org.bson.Document
import org.graalvm.polyglot.Value

class AggregationCursor<out T> internal constructor(private val cursor: Value, private val context: MongoShellContext) : Cursor<T> {
    override fun hasNext(): Boolean {
        return cursor.invokeMember("hasNext").asBoolean()
    }

    override fun next(): T {
        if (!hasNext()) throw NoSuchElementException()
        return context.extract(cursor.invokeMember("next")).value as T
    }

    override fun toReplString(): String {
        return context.extract(cursor.invokeMember("toReplString")).toReplString()
    }
}